### PR TITLE
Ros1 development

### PR DESCRIPTION
**Pull request by rafal-gorecki**

### CODE REVIEW

File: demo/compose.desktop.yaml [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    environment:
+      - DISPLAY
+      - QT_X11_NO_MITSHM=1
+      - NVIDIA_VISIBLE_DEVICES=all
+      - NVIDIA_DRIVER_CAPABILITIES=compute,utility,display,video
+    volumes:
+      - /tmp/.X11-unix:/tmp/.X11-unix:rw
+      - ./params:/ros_ws/src/zed-ros-wrapper/zed_wrapper/params
+    privileged: true
+    command: roslaunch zed_wrapper <camera_model>.launch
```
Comment (KmakD) [inline demo/compose.desktop.yaml]:
```suggestion
    command: roslaunch zed_wrapper ${CAMERA_MODEL}.launch
```
This way you don't need to modify compose each time you change camera you just:
```
export CAMERA_MODEL=<camera_model>
docker compose up
```
And if someone is always using same camera just add export to `.bashrc` 
Same applies in other compose files

File: demo/compose.desktop.yaml [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    tty: true
+    environment:
+      - DISPLAY
+      - QT_X11_NO_MITSHM=1
+      - NVIDIA_VISIBLE_DEVICES=all
+      - NVIDIA_DRIVER_CAPABILITIES=compute,utility,display,video
+    volumes:
+      - /tmp/.X11-unix:/tmp/.X11-unix:rw
+      - ./params:/ros_ws/src/zed-ros-wrapper/zed_wrapper/params
+    privileged: true
```
Comment (KmakD) [inline demo/compose.desktop.yaml]: Generally using privileged is not advised, is it really needed in this case? Can't we just specify correct device and volumes for zed to work? same applies to other compose files

File: demo/compose.desktop.yaml [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    command: roslaunch zed_wrapper <camera_model>.launch
+
+  rviz:
+    image: husarion/rviz:noetic
+    network_mode: "host"
```
Comment (KmakD) [inline demo/compose.desktop.yaml]:
```suggestion
    image: husarion/rviz:noetic
    container_name: rviz
    network_mode: host
```
same applies to other compose files

File: .github/workflows/build-docker-image.yaml [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+      - name: Build L4T32 Docker Image
+        uses: docker/build-push-action@v3
+        with:
+          file: Dockerfile.jetson_l4t32
+          platforms: linux/arm64
+          push: true
+          tags: |
+            husarion/zed-jetson:melodic-nightly
+            husarion/zed-jetson:melodic
+
```
Comment (KmakD) [inline .github/workflows/build-docker-image.yaml]: This could be done by adding it as another matrix entry

File: .gitignore
Comment (KmakD) [inline .gitignore:1]: That file is very random, consider using .gitignore same or similar to this one: https://github.com/husarion/panther-docker/blob/ros1/.gitignore

File: Dockerfile.desktop [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+### BUILD IMAGE ###
+FROM nvidia/cuda:${CUDA_MAJOR}.${CUDA_MINOR}.${CUDA_PATCH}-devel-ubuntu${UBUNTU_RELEASE_YEAR}.04 AS build
+
+ARG UBUNTU_RELEASE_YEAR=20
+ARG CUDA_MAJOR=11
+ARG CUDA_MINOR=7
+ARG CUDA_PATCH=1
+ARG ZED_SDK_MAJOR=4
+ARG ZED_SDK_MINOR=0
+ARG ROS_DISTRO=noetic
```
Comment (KmakD) [inline Dockerfile.desktop]:
```suggestion
ARG UBUNTU_RELEASE_YEAR=20
ARG CUDA_MAJOR=11
ARG CUDA_MINOR=7
ARG CUDA_PATCH=1
ARG ZED_SDK_MAJOR=4
ARG ZED_SDK_MINOR=0
ARG ROS_DISTRO=noetic

### BUILD IMAGE ###
FROM nvidia/cuda:${CUDA_MAJOR}.${CUDA_MINOR}.${CUDA_PATCH}-devel-ubuntu${UBUNTU_RELEASE_YEAR}.04 AS build

ARG UBUNTU_RELEASE_YEAR
ARG CUDA_MAJOR
ARG CUDA_MINOR
ARG CUDA_PATCH
ARG ZED_SDK_MAJOR
ARG ZED_SDK_MINOR
ARG ROS_DISTRO
```
If you have global ARGs defined there is no need to reassign default values

File: Dockerfile.desktop
Comment (KmakD) [inline Dockerfile.desktop:1]: Reducing number of steps for runtime stage should also reduce image size. You should be able to reduce it to single step.
Also, removing unused dependencies and packages may help:
```
    apt-get remove -y \
        python3-rosdep \
        python3-vcstool \
        git && \ 
    apt-get autoremove -y && \
    apt-get clean && \
    rm -rf /var/lib/apt/lists/*
```
for reference see: https://github.com/husarion/rosbot-xl-docker/blob/main/Dockerfile.hardware

File: Dockerfile.desktop [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+
+COPY --from=build /ros_ws /ros_ws
+
+ARG UBUNTU_RELEASE_YEAR=20
+ARG CUDA_MAJOR=11
+ARG CUDA_MINOR=7
+ARG CUDA_PATCH=1
+ARG ZED_SDK_MAJOR=4
+ARG ZED_SDK_MINOR=0
+ARG ROS_DISTRO=noetic
```
Comment (KmakD) [inline Dockerfile.desktop]:
```suggestion
ARG UBUNTU_RELEASE_YEAR
ARG CUDA_MAJOR
ARG CUDA_MINOR
ARG CUDA_PATCH
ARG ZED_SDK_MAJOR
ARG ZED_SDK_MINOR
ARG ROS_DISTRO
```

File: Dockerfile.jetson [hunk 7]
```diff
@@ -0,0 +1,136 @@
+ARG L4T_MAJOR=35
+ARG L4T_MINOR=3
+ARG CUDA_MAJOR=11
+ARG CUDA_MINOR=4
+ARG CUDA_PATCH=19
+ARG ZED_SDK_MAJOR=4
+ARG ZED_SDK_MINOR=0
```
Comment (KmakD) [inline Dockerfile.jetson:7]:
```suggestion
ARG ZED_SDK_MINOR=0
ARG ROS_DISTRO=noetic
```
Lets make this similar between dockerfiles

File: Dockerfile.jetson [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+### BUILD IMAGE ###
+FROM nvcr.io/nvidia/l4t-cuda:${CUDA_MAJOR}.${CUDA_MINOR}.${CUDA_PATCH}-devel AS build
+
+ARG L4T_MAJOR=35
+ARG L4T_MINOR=3
+ARG CUDA_MAJOR=11
+ARG CUDA_MINOR=4
+ARG CUDA_PATCH=19
+ARG ZED_SDK_MAJOR=4
+ARG ZED_SDK_MINOR=0
```
Comment (KmakD) [inline Dockerfile.jetson]:
```suggestion
ARG L4T_MAJOR
ARG L4T_MINOR
ARG CUDA_MAJOR
ARG CUDA_MINOR
ARG CUDA_PATCH
ARG ZED_SDK_MAJOR
ARG ZED_SDK_MINOR
```

File: Dockerfile.jetson [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+### RUNTIME IMAGE ### # Reduction from 6.3GB to 4.1GB
+FROM nvcr.io/nvidia/l4t-cuda:${CUDA_MAJOR}.${CUDA_MINOR}.${CUDA_PATCH}-runtime AS runtime
+
+ARG L4T_MAJOR=35
+ARG L4T_MINOR=3
+ARG CUDA_MAJOR=11
+ARG CUDA_MINOR=4
+ARG CUDA_PATCH=19
+ARG ZED_SDK_MAJOR=4
+ARG ZED_SDK_MINOR=0
```
Comment (KmakD) [inline Dockerfile.jetson]:
```suggestion
ARG L4T_MAJOR
ARG L4T_MINOR
ARG CUDA_MAJOR
ARG CUDA_MINOR
ARG CUDA_PATCH
ARG ZED_SDK_MAJOR
ARG ZED_SDK_MINOR
```

File: Dockerfile.jetson
Comment (KmakD) [inline Dockerfile.jetson:1]: same as `Dockerfile.desktop` image size could be further reduced

File: Dockerfile.jetson_l4t32 [hunk 10]
```diff
@@ -0,0 +1,53 @@
+ARG L4T_MAJOR=32
+ARG L4T_MINOR=7
+ARG ZED_SDK_MAJOR=3
+ARG ZED_SDK_MINOR=8
```
Comment (KmakD) [inline Dockerfile.jetson_l4t32]:
```suggestion
ARG ZED_SDK_MINOR=8
ROS_DISTRO=melodic
```
same as before

File: Dockerfile.jetson_l4t32 [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+
+FROM stereolabs/zed:${ZED_SDK_MAJOR}.${ZED_SDK_MINOR}-devel-l4t-r${L4T_MAJOR}.${L4T_MINOR} AS build
+
+ARG L4T_MAJOR=32
+ARG L4T_MINOR=7
+ARG ZED_SDK_MAJOR=3
+ARG ZED_SDK_MINOR=8
```
Comment (KmakD) [inline Dockerfile.jetson_l4t32]:
```suggestion
ARG L4T_MAJOR
ARG L4T_MINOR
ARG ZED_SDK_MAJOR
ARG ZED_SDK_MINOR
```

File: Dockerfile.jetson_l4t32
Comment (KmakD) [inline Dockerfile.jetson_l4t32:1]: I'm a bit confused about this one. You have only one stage that is called `build`. Then you don't build packages directly in dockerfile but call it in entrypoint file. Why can't we do it the same way as in previous dockerfiles? I really don't see a point in doing it this way.

File: demo/compose.jetson_l4t32.yaml [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+    volumes:
+      - /tmp/.X11-unix:/tmp/.X11-unix:rw
+      - ./default.rviz:/default.rviz
+    environment:
+      - DISPLAY
+      - LIBGL_ALWAYS_SOFTWARE=1
+    command: rosrun rviz rviz -d /default.rviz
+
+volumes:
+  workspace:
```
Comment (KmakD) [inline demo/compose.jetson_l4t32.yaml]: Why do we need that volume?

File: ros_entrypoint_l4t32.sh [hunk 13]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
+set -e
+
+# setup ros environment
+source "/opt/ros/$ROS_DISTRO/setup.bash"
+catkin_make -DCMAKE_BUILD_TYPE=Release || true
```
Comment (KmakD) [inline ros_entrypoint_l4t32.sh]: As mentioned before i dont think that is a good way of building ros packages

File: README.md [hunk 14]
```diff
@@ -1,2 +1,47 @@
-# zed2-docker
-Dockerized Stereolabs ZED2 and ZED2i package
+# zed-docker
+
+The repository contains a GitHub Actions workflow for auto-deployment of built Docker image to https://hub.docker.com/r/husarion/zed_desktop & https://hub.docker.com/r/husarion/zed_jetson repository.
```
Comment (KmakD) [inline README.md]:
```suggestion
The repository contains a GitHub Actions workflow for auto-deployment of built Docker images to [husarion/zed_desktop](https://hub.docker.com/r/husarion/zed-desktop) and [husarion/zed_jetson](https://hub.docker.com/r/husarion/zed-jetson) Docker Hub repositories.
```

File: README.md [hunk 15]
```diff
@@ -1,2 +1,47 @@
...
+
+## Docker image usage
+
+Available repos: 
+- **`zed-desktop:noetic`** for desktop platform with CUDA (tested on platform with 11.7).
+- **`zed-jetson:noetic`** for Jetson platform currently support - **Jetson Xavier, Orin AGX/NX/Nano, CUDA 11.4** (tested on Xavier AGX).
```
Comment (KmakD) [inline README.md]:
```suggestion
- **`zed-jetson:noetic`** for Jetson platform currently supported - **Jetson Xavier, Orin AGX/NX/Nano, CUDA 11.4** (tested on Xavier AGX).
```

File: README.md [hunk 16]
```diff
@@ -1,2 +1,47 @@
...
+
+```bash
+docker run --runtime nvidia -it --privileged --ipc=host --pid=host -e DISPLAY \
+  -v /dev/shm:/dev/shm -v /tmp/.X11-unix/:/tmp/.X11-unix \
+  -v /tmp/zed_ai/:/usr/local/zed/resources/ \
+  <image_name> \
+  roslaunch zed_wrapper <camera_model>.launch
+```
+
+## Docker Compose
```
Comment (KmakD) [inline README.md]: Personally I would put that before development section. Also, consider renaming it to something like `Running demo` or `Running demo with Docker compose`

**Review by KmakD — CHANGES_REQUESTED** [review]
Add also .dockerignore file similar to this one: https://github.com/husarion/panther-docker/blob/ros1/.dockerignore

File: demo/compose.desktop.yaml (hunk 1 above)
Comment (rafal-gorecki) [reply demo/compose.desktop.yaml]: Good idea! Thanks

**Pull request merged**